packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+            uint256 amount = moxieBondingCurve.buyShares(
+                _subjects[i],
+                _depositAmounts[i],
+                _minReturnAmountsAfterFee[i]
+            );
+            (, unlockTimeInSec_) = _createLock(
+                _subjects[i],
+                amount,
+                _lockPeriodInSec,
+                msg.sender,
+                true
+            );
+            amounts_[i] = amount;
+        }
+    }
+
+    /**
+     * External function to buy & lock multiple tokens.
+     * @param _subjects Subject addresses for which tokens are being bought & deposited.
+     * @param _depositAmounts Amounts of moxie tokens getting deposited.
+     * @param _minReturnAmountsAfterFee Slippage settings which determine minimum amounts of tokens after fee.
+     * @param _lockPeriodInSec Lock periods for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @return amounts_ Amounts of tokens bought & locked.
+     * @return unlockTimeInSec_ Unlock times for the locks.
+     */
+    function buyAndLockMultipleFor(
+        address[] memory _subjects,
+        uint256[] memory _depositAmounts,
+        uint256[] memory _minReturnAmountsAfterFee,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256[] memory amounts_, uint256 unlockTimeInSec_)
+    {
+        if (
+            _subjects.length != _depositAmounts.length ||
+            _subjects.length != _minReturnAmountsAfterFee.length
+        ) {
+            revert Staking_InvalidInputLength();
+        }
+
+        uint256 totalDepositAmount = 0;
+
+        for (uint256 i = 0; i < _subjects.length; i++) {
+            totalDepositAmount += _depositAmounts[i];
+        }
+
+        moxieToken.safeTransferFrom(
+            msg.sender,
+            address(this),
+            totalDepositAmount
+        );
+        moxieToken.approve(address(moxieBondingCurve), totalDepositAmount);
+
+        amounts_ = new uint256[](_subjects.length);
+
+        for (uint256 i = 0; i < _subjects.length; i++) {
+            amounts_[i] = moxieBondingCurve.buyShares(
+                _subjects[i],
+                _depositAmounts[i],
+                _minReturnAmountsAfterFee[i]
+            );
+            (, unlockTimeInSec_) = _createLock(
+                _subjects[i],
+                amounts_[i],
+                _lockPeriodInSec,
+                _beneficiary,
+                true
+            );
+        }
+    }
+
+    /**
+     * External function to withdraw locked tokens.
+     * @param _subject Subject address for which tokens are being withdrawn.
+     * @param _indexes Indexes of the locks to be withdrawn.
+     * @return totalAmount_ Total amount of tokens withdrawn.
+     */
+    function withdraw(
+        address _subject,
+        uint256[] memory _indexes
+    ) external returns (uint256 totalAmount_) {
```

Comment:
May be we should add Pausable modifier and only allow withdraw when not paused

Reply:
Yes we should add @sharathkrml